diff --git a/fn/-z4h-cmd-ssh b/fn/-z4h-cmd-ssh
index b7a0732..025b29c 100644
--- a/fn/-z4h-cmd-ssh
+++ b/fn/-z4h-cmd-ssh
@@ -186,7 +186,7 @@ if (( must_passthrough || !z4h_ssh_enable )); then
 fi
 
 if (( $#z4h_retrieve_history )); then
-  local local_hist_tmp=$Z4H/tmp/ssh-history.tmp.$sysparams[pid]
+  local local_hist_tmp={Z4H_TMP:-$Z4H/tmp}/ssh-history.tmp.$sysparams[pid]
   z4h_ssh_retrieve_files[\$HISTFILE]=$local_hist_tmp
   zf_rm -f -- $local_hist_tmp || return
 else
@@ -245,9 +245,9 @@ fi
 
 local tmpdir
 if (( $+commands[mktemp] )); then
-  tmpdir=$(command mktemp -d -- $Z4H/tmp/ssh.XXXXXXXXXX) || return
+  tmpdir=$(command mktemp -d -- {Z4H_TMP:-$Z4H/tmp}/ssh.XXXXXXXXXX) || return
 else
-  tmpdir=$Z4H/tmp/ssh.tmp.$sysparams[pid]
+  tmpdir={Z4H_TMP:-$Z4H/tmp}/ssh.tmp.$sysparams[pid]
   zf_rm -rf -- $tmpdir || return
   zf_mkdir -- $tmpdir || return
 fi
@@ -368,9 +368,9 @@ local -i pid=$sysparams[pid]
    return $pipestatus[1]
 } 3>&1 | {
   if (( $+commands[mktemp] )); then
-    tmpdir=$(command mktemp -d -- $Z4H/tmp/ssh.XXXXXXXXXX) || return
+    tmpdir=$(command mktemp -d -- {Z4H_TMP:-$Z4H/tmp}/ssh.XXXXXXXXXX) || return
   else
-    tmpdir=$Z4H/tmp/ssh.tmp.$sysparams[pid]
+    tmpdir={Z4H_TMP:-$Z4H/tmp}/ssh.tmp.$sysparams[pid]
     zf_rm -rf -- $tmpdir || return
     zf_mkdir -- $tmpdir || return
   fi
diff --git a/fn/-z4h-cmd-update b/fn/-z4h-cmd-update
index 402a9b8..b12105a 100644
--- a/fn/-z4h-cmd-update
+++ b/fn/-z4h-cmd-update
@@ -25,7 +25,7 @@ local new=$Z4H.new.$$
     (( $? )) && "exit" "1"
     "builtin" "emulate" "zsh" "-o" "no_aliases"
     [[ $Z4H == '${(q)new}' ]] || exit 0
-    print -n >$Z4H/tmp/update-successful' || return
+    print -n >{Z4H_TMP:-$Z4H/tmp}/update-successful' || return
 
   if [[ ! -e $new/tmp/update-successful ]]; then
     local home=~
diff --git a/fn/-z4h-compinit b/fn/-z4h-compinit
index 3abea05..a0853d5 100644
--- a/fn/-z4h-compinit
+++ b/fn/-z4h-compinit
@@ -50,7 +50,7 @@ if [[ -r $dump ]] &&
     (( stat[2] == stat[1] + 1 )); then
   -z4h-compinit-impl -C -d $dump
 else
-  local tmp=$Z4H/tmp/zcompdump.$sysparams[pid]
+  local tmp={Z4H_TMP:-$Z4H/tmp}/zcompdump.$sysparams[pid]
   zf_rm -f -- $dump $dump.zwc $tmp $tmp.2
   -z4h-compinit-impl -C -d $tmp
   { print -rn -- $real_sig; <$tmp } >$tmp.2
diff --git a/fn/-z4h-install-many b/fn/-z4h-install-many
index 7302f48..0ce8005 100644
--- a/fn/-z4h-install-many
+++ b/fn/-z4h-install-many
@@ -86,9 +86,9 @@ local success
   fi
 
   if (( $+commands[mktemp] )); then
-    success="$(command mktemp -- $Z4H/tmp/install-many.XXXXXXXXXX)" || return
+    success="$(command mktemp -- {Z4H_TMP:-$Z4H/tmp}/install-many.XXXXXXXXXX)" || return
   else
-    success=$Z4H/tmp/install-many.$$
+    success={Z4H_TMP:-$Z4H/tmp}/install-many.$$
     zf_rm -rf -- $success   || return
     print -n >$success      || return
   fi
